# FOR LOOPS
Loops are used for sequential traversal. For traversing list, string, tuples etc.

In [29]:
veggies = ["Potato", "Brinjal", "Ladyfinger", "Tomato", "Cucumber", "Onion"]

for _ in veggies:
    print(_)

Potato
Brinjal
Ladyfinger
Tomato
Cucumber
Onion


Printing numbers in a Single Line

In [50]:
for i in range(1, 6):
    print(i, end=" ")

1 2 3 4 5 

In [30]:
name = "SanaUrRehmanArain"
for _ in name:
    print(_)

S
a
n
a
U
r
R
e
h
m
a
n
A
r
a
i
n


In [1]:
countries = ["China", "Portugal", "Spain", "Germany", "UK", "U.S.A", "Pakistan", "India", "Sweden", "Norway"]

for i in range(len(countries)):
    print(i, countries[i])

0 China
1 Portugal
2 Spain
3 Germany
4 UK
5 U.S.A
6 Pakistan
7 India
8 Sweden
9 Norway


**How It Works**
1. `len(countries)` → finds how many items are in the list.

   * Here: `len(countries)` = **10**

2. `range(len(countries))` → generates numbers from **0** to **9**.

3. For each `i` in that range:

   * `countries[i]` accesses the country at index i.

   * Prints `i` (index) and `countries[i]` (country name).

**Visual**

| `i` | `countries[i]` |
| --- | -------------- |
| 0   | China          |
| 1   | Portugal       |
| 2   | Spain          |
| 3   | Germany        |
| 4   | UK             |
| 5   | U.S.A          |
| 6   | Pakistan       |
| 7   | India          |
| 8   | Sweden         |
| 9   | Norway         |

In [2]:
for i, j in enumerate(countries):
    print(i)
    print(j)

0
China
1
Portugal
2
Spain
3
Germany
4
UK
5
U.S.A
6
Pakistan
7
India
8
Sweden
9
Norway


**How It Works**
1. `enumerate(countries)` returns **pairs** → `(index, value)`.

2. For each `(i, j)`:

   - `i` → the index of the country.

   - `j` → the country name.

3. Prints the index (`i`) **on one line** and the country (`j`) **on the next**.

**Visual**
| `i` | `j` (country) |
| --- | ------------- |
| 0   | China         |
| 1   | Portugal      |
| 2   | Spain         |
| 3   | Germany       |
| 4   | UK            |
| 5   | U.S.A         |
| 6   | Pakistan      |
| 7   | India         |
| 8   | Sweden        |
| 9   | Norway        |

**Key Difference:**

- `range(len(...))` → You manually get indexes, then access the value.

- `enumerate(...)` → Python automatically gives you both index and value.

In [3]:
dict = {
    "name": "John",
    "age": 70,
    "profession": "mechanic",
    "education": "diploma"
}

for key, value in dict.items():
    print(key)
    print(value)

name
John
age
70
profession
mechanic
education
diploma


## Range
Range functions returns a sequence of numbers, starting from 0 by default, and increments by 1 (by default), and stops before a specified number.

In [31]:
for _ in range(3): # starts from 0, ends at 2, range(stop)
    print(_)

0
1
2


In [32]:
for _ in range(1, 5): # starts from 1, end at 4, range(start, stop)
    print(_)

1
2
3
4


In [37]:
for _ in range(2, 10, 2): # step size 2, range(start, stop, step)
    print(_)

2
4
6
8


In [38]:
# printing even numbers
for i in range(2, 10, 2):
    print(i)

2
4
6
8


In [39]:
# printing odd numbers
for i in range (1, 10, 2):
    print(i)

1
3
5
7
9


##### Print numbers from 1 to 100.

In [40]:
for i in range(1, 101):
    print(i)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100


##### Print numbers from 100 to 1.

In [23]:
for i in range(100, 0, -1):
    print(i)

100
99
98
97
96
95
94
93
92
91
90
89
88
87
86
85
84
83
82
81
80
79
78
77
76
75
74
73
72
71
70
69
68
67
66
65
64
63
62
61
60
59
58
57
56
55
54
53
52
51
50
49
48
47
46
45
44
43
42
41
40
39
38
37
36
35
34
33
32
31
30
29
28
27
26
25
24
23
22
21
20
19
18
17
16
15
14
13
12
11
10
9
8
7
6
5
4
3
2
1


##### Print the multiplication table of a number n.

In [41]:
n = int(input("enter number: "))

for i in range (1, 11):
    print(f"{n} x {i} = {n * i}")

2 x 1 = 2
2 x 2 = 4
2 x 3 = 6
2 x 4 = 8
2 x 5 = 10
2 x 6 = 12
2 x 7 = 14
2 x 8 = 16
2 x 9 = 18
2 x 10 = 20


##### Print the elements of the following list using a loop:
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

In [27]:
numbers = [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
for num in numbers:
    print(num)

1
4
9
16
25
36
49
64
81
100


##### Search for a number x in this tuple using loop:
(1, 4, 9, 16, 25, 36, 49, 64, 81, 100)

In [28]:
numbers = (1, 4, 9, 16, 25, 36, 49, 64, 81, 100)
x = 36

for i in range(len(numbers)):
    if numbers[i] == x:
        print("Found at index:", i)
        break  # Stops the loop once the number is found

Found at index: 5


## Pass Statement
Pass is a null statement that does nothing. It is used as a placeholder for future code.

In [44]:
for i in range (5):
    pass # now it will not throw an arror as there should be something in the loop.

print("I will use/fil 'for loop' in future")

I will use/fil 'for loop' in future


##### WAP to find the factorial of first n natural numbers. 

In [49]:
# 1 * 2 * 3 * 4 * 5
n = 5

fact = 1
for i in range(1, n + 1):
    fact *= i

print("Factorial = ", fact)

# If we used range(1, n), it would generate:
# 1, 2, 3, 4  (excluding `5`, which is incorrect for factorial)
# Thus, n + 1 ensures the loop runs from 1 to n, covering all necessary numbers for factorial calculation.

Factorial =  120


##### WAP to find the sum of first n natural numbers.

In [45]:
# 1 + 2 + 3 + 4 + 5
n = 5

sum = 0
for i in range(1, n + 1):
    sum += i

print("Total sum = ", sum)

Total sum =  15


In [10]:
a = [0, 1, 2, 3, 4, 5] # list
b = (0, 1, 2, 3, 4, 5) # tuple
c = {0, 1, 2, 3, 4, 5} # set
d = "012345" 		   # string
e = {				   # dictionary
	"name": "max",
	"age": 20,
	"city": "Braga"
}

print(1 in a) # to check whether 1 is present in the list, output True
print(3 in c) # to check whether 3 is present in the set, output True
print("name" in e) # to check whether "name" is present in dict, output True
print(100 in b) # to check whether 100 is present in tuple, output False
print("4" in d) # # to check whether 4 is present in the string output True

True
True
True
False
True


In [11]:
for _ in a: # iterate through all the items in the list
	print(_)

0
1
2
3
4
5


In [12]:
for x in e.keys(): # return all the keys in the dict
	print(x)

name
age
city


In [13]:
for x in e.values(): # return all the values in the dict
	print(x)

max
20
Braga


In [15]:
for x in e.items(): # return all the item (keys & values) in the dict
	print(x) # output: ('name', 'max')('age', 20)('city', 'Braga')

('name', 'max')
('age', 20)
('city', 'Braga')


In [16]:
for key, value in e.items():
	print(key, " ", value) # output: name   max  age   20  city   Braga

name   max
age   20
city   Braga


In [17]:
for _ in range(6): # output start from 0 to 5
	print(_)

0
1
2
3
4
5


In [18]:
for _ in range (2, 6): # output start from 2 to 5
	print(_)

2
3
4
5


In [19]:
for _ in range (10, 60, 10): # output: 10, 20, 30, 40, 50
	print(_)

10
20
30
40
50


Printing Meow three times can be tideous.

In [14]:
print("Meow\n" * 3, end = "")

Meow
Meow
Meow


We can write easily with for loops.

In [6]:
# Method 1
for i in [0, 1, 2]:
    print("Meow")

Meow
Meow
Meow


In [9]:
# Method 2
# for _ in range(3):, we can also use this, becuase we will not use this variable 
# again and it is only needed for this loop and we don't care about it's name.
for i in range(3): 
    print("Meow") 

Meow
Meow
Meow


In [3]:
def main():
    names = ["Mario", "Daisy", "Luigi", "Yoshi"]

    # Method 1
    for i in range(len(names)):
            print(write_letter(names[i], "Princess Peach"))

def write_letter(receiver, sender):
    return f"""
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
        Dear {receiver},

        You are cordially invited to a ball at
        Peach's Castle this evening, 7:00 PM.

        Sincerely,
        {sender}
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
    """
main()


    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
        Dear Mario,

        You are cordially invited to a ball at
        Peach's Castle this evening, 7:00 PM.

        Sincerely,
        Princess Peach
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
    

    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
        Dear Daisy,

        You are cordially invited to a ball at
        Peach's Castle this evening, 7:00 PM.

        Sincerely,
        Princess Peach
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
    

    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
        Dear Luigi,

        You are cordially invited to a ball at
        Peach's Castle this evening, 7:00 PM.

        Sincerely,
        Princess Peach
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
    

    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
        Dear Yoshi,

        You are cordially invited to a ball at
        Peach's Castle this evening, 7:00 PM.

        Sincerely,
        Princess Pe

The grocery owner wants to discount all inventory items by 30%. We can do this easily by using for loops.

In [4]:
def main():
    names = ["Mario", "Daisy", "Luigi", "Yoshi"]   
    
    # method 2, better version
    for name in names:
        print(write_letter(name, "Princess Peach"))

def write_letter(receiver, sender):
    return f"""
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
        Dear {receiver},

        You are cordially invited to a ball at
        Peach's Castle this evening, 7:00 PM.

        Sincerely,
        {sender}
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
    """
main()


    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
        Dear Mario,

        You are cordially invited to a ball at
        Peach's Castle this evening, 7:00 PM.

        Sincerely,
        Princess Peach
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
    

    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
        Dear Daisy,

        You are cordially invited to a ball at
        Peach's Castle this evening, 7:00 PM.

        Sincerely,
        Princess Peach
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
    

    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
        Dear Luigi,

        You are cordially invited to a ball at
        Peach's Castle this evening, 7:00 PM.

        Sincerely,
        Princess Peach
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
    

    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
        Dear Yoshi,

        You are cordially invited to a ball at
        Peach's Castle this evening, 7:00 PM.

        Sincerely,
        Princess Pe

In [1]:
prices = [
    51.79, 39.37, 17.30, 38.42, 12.41, 19.38, 52.23, 43.92, 11.88, 7.49,
    46.02, 79.85, 50.51, 97.94, 38.57, 76.62, 53.61, 45.44, 75.90, 21.01, 
    6.66, 71.08, 86.49, 70.11, 98.59, 9.52, 72.78, 91.21, 32.01, 28.35
]
discounted_prices = []   # make a empty list to store discounted prices

for price in prices: # take each item from the list prices one by one and put it in the Jar(iteration / loop_variable price). iterate through each item until the list is finished.
    discounted_prices.append(round(price * .7, 2))  # multiply them with .7 and append it to the list named discounted_prices.

print(discounted_prices)

[36.25, 27.56, 12.11, 26.89, 8.69, 13.57, 36.56, 30.74, 8.32, 5.24, 32.21, 55.89, 35.36, 68.56, 27.0, 53.63, 37.53, 31.81, 53.13, 14.71, 4.66, 49.76, 60.54, 49.08, 69.01, 6.66, 50.95, 63.85, 22.41, 19.84]


Python reaches the loop's header for the first time.

It creates the iteration variable price and assigns it the first item of the prices list.

Then Python executes the code in the body of the loop.

In this case, the body of the loop contains a single line of code, but there could be more.

Since the iteration variable price takes the value 51.79 during the first iteration, the value 36.25

is appended to the discounted prices list following the execution of the code in the body of the loop.

Python returns to the header.

At that point, the iteration variable is reassigned.

The second item in the prices list.

Then Python executes the code in the body of the loop again.

Since the iteration variable takes the value 39.37 during the second iteration, the value 27.56 is

appended to the discounted price list.

Python then returns to the header reassigns the iteration variable and executes the code in the body

of the loop once again.

This process continues until the iteration variable has been assigned every price in the prices list.

When Python runs out of item, the loop execution is complete.

In [20]:
waypoints = [
    "Granite Falls Crossing",
    "Starlight Meadow",
    "Cedar Creek Junction",
    "Evergreen Grove",
    "Pine Ridge Summit"
]

tasks = [
    "Drink water",
    "Eat a snack",
    "Inventory supplies",
    "Take pictures"
]

In [21]:
for waypoint in waypoints:
    print(f"Arrived at {waypoint}, completing tasks:")
    print(f"\t{tasks[0]}")
    print(f"\t{tasks[1]}")
    print(f"\t{tasks[2]}")
    print(f"\t{tasks[3]}")
    print(f"Leaving {waypoint}.\n")

Arrived at Granite Falls Crossing, completing tasks:
	Drink water
	Eat a snack
	Inventory supplies
	Take pictures
Leaving Granite Falls Crossing.

Arrived at Starlight Meadow, completing tasks:
	Drink water
	Eat a snack
	Inventory supplies
	Take pictures
Leaving Starlight Meadow.

Arrived at Cedar Creek Junction, completing tasks:
	Drink water
	Eat a snack
	Inventory supplies
	Take pictures
Leaving Cedar Creek Junction.

Arrived at Evergreen Grove, completing tasks:
	Drink water
	Eat a snack
	Inventory supplies
	Take pictures
Leaving Evergreen Grove.

Arrived at Pine Ridge Summit, completing tasks:
	Drink water
	Eat a snack
	Inventory supplies
	Take pictures
Leaving Pine Ridge Summit.



### we can make a better version by creating a nested loop. 

In [22]:
for waypoint in waypoints:
    print(f"Arrived at {waypoint}, completing tasks:")
    for task in tasks:
        print(f"\t{task}")
    print(f"Leaving {waypoint}.\n")

Arrived at Granite Falls Crossing, completing tasks:
	Drink water
	Eat a snack
	Inventory supplies
	Take pictures
Leaving Granite Falls Crossing.

Arrived at Starlight Meadow, completing tasks:
	Drink water
	Eat a snack
	Inventory supplies
	Take pictures
Leaving Starlight Meadow.

Arrived at Cedar Creek Junction, completing tasks:
	Drink water
	Eat a snack
	Inventory supplies
	Take pictures
Leaving Cedar Creek Junction.

Arrived at Evergreen Grove, completing tasks:
	Drink water
	Eat a snack
	Inventory supplies
	Take pictures
Leaving Evergreen Grove.

Arrived at Pine Ridge Summit, completing tasks:
	Drink water
	Eat a snack
	Inventory supplies
	Take pictures
Leaving Pine Ridge Summit.

